### ISSUE TITLE
Resolve for side-effects only

### ISSUE THREAD

**Issue by sliekens** [angular/angular#25428]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[ ] Bug report  
[ ] Performance issue
[x] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
[ ] Other... Please describe:
</code></pre>

## Current behavior

Sometimes it's convenient to use a Resolve service just to make side-effects happen before a route is activated.

```typescript
// route config
{
  path: 'startup',
  pathMatch: 'full',
  component: StartupComponent,
  resolve: {
    dowork:  DoImportantStuff
  }
 }

// resolve service that does stuff before the component is initialized
@Injectable()
export class DoImportantStuff implements Resolve<void> {
    resolve(route: ActivatedRouteSnapshot, state: RouterStateSnapshot): Observable<void> | Promise<void> | void {
        // do important stuff
        // example: load user data into session storage
        // then return void
    }
}
```

This code is really easy to write but at the same time it feels awkward to use a Resolve service for side-effects only.
- It implements a generic interface with `void` as the type parameter: weird!
- The routeconfig needs a name for the void result: weird!
- The route snapshot will have a data property that's always undefined: weird!
`route.snapshot.data.dowork === undefined // always`

## Expected behavior

Add an API that's similar to Resolve but for doing background work that doesn't return a value.

Let's call it route initializers. Not to be confused with app initializers because these babies are only used when the route is activated.

Oh and they run after `CanActivate`.

```typescript
// new interface
interface OnRouteActivated {
  init(route: ActivatedRouteSnapshot, state: RouterStateSnapshot): Observable<void> | Promise<void> | void;
}

// extend Route with new interface
interface Route {
  initializers?: OnRouteActivated[]
}
```

This is essentially still the same as Resolve but without upsetting CQS purists.

## What is the motivation / use case for changing the behavior?

Sometimes it's difficult to fit background work in a component's `ngOnInit` method
- especially if it involves doing things in parallel.
- or because you need the same initialization code for multiple routes.

A Resolve service seems like a good solution to these problems, but the code feels awkward because a Resolve is expected to return a result.

FYI this is what it would look like in the new world:

```ts
// route initializer service that does stuff before the component is initialized
@Injectable()
export class DoImportantStuff implements OnRouteActivated {
    init(route: ActivatedRouteSnapshot, state: RouterStateSnapshot): Observable<void> | Promise<void> | void {
        // do important stuff
        // example: load user data into session storage
        // then return void (or a void promise/observable)
    }
}

// route config
{
  path: 'startup',
  pathMatch: 'full',
  component: StartupComponent,
  // initializers don't have a key because they don't return data 
  // there can be more than one, order shouldn't matter, preferably they run concurrently
  initialize: [ DoImportantStuff ]
 }
```

**Comment by th0r** [comment]
Let me add my usecase. I think that adding a separate resolver class stored in a separate file just for fetching a data (and in most cases it's just a single call to a service method) is a perfect example of overengineering. Recently I came up with an interesting idea - use page component classes as route resolvers. Here is an example:
```tr
interface Post {
  id: number;
  title: string;
}

const routes: Route[] = [
  {
    path: 'post/:id',
    resolve: {
      post: PostPageComponent
    }
  }
];

@Component({
  selector: 'app-post-page',
  template: 'Title: {{ post.title }}'
})
class PostPageComponent implements OnInit, Resolve<Post> {
  post: Post;

  constructor(
    private posts: PostsService,
    private route: ActivatedRoute
  ) {}

  async resolve(route: ActivatedRouteSnapshot): Promise<Post> {
    return this.posts.get(route.params.id);
  }

  ngOnInit() {
    this.post = this.route.snapshot.data.post;
  }
}

@Injectable()
class PostsService {
  async get(id: Post['id']): Promise<Post> {
    // ...fetch post...
    return {id: 1, title: 'Foo'};
  }
}
```
So I just use `PostPageComponent` as resolver for post page data and it works great except one small nuance: I have to provide a name for resolved resource in route configuration (`post` key in `resolve` object) but I don't need it because I want `PostComponent` to fetch all the necessary data and put it straight into the `route.snapshot.data` object. Of course I can rename `post` key into `data` but in this case accessing it will look like `route.snapshot.data.data` and it bothers me a little.

So what I propose is to allow `Route['resolve']` property to be a single value / observable / promise. In this case it's resolved value will be stored right into `route.snapshot.data`. I think this will solve your problem as well, right @StevenLiekens?

**Comment by StevenLiekens** [comment]
@th0r I think I understand what you're saying but it wouldn't work for routes that also have static data.

Example: `title` would get lost if the Resolve replaces the whole data object 
```ts
{
  path: 'posts',
  component: PostPageComponent,
  data: { title: 'Posts List' },
  resolve: ResolvePostPageData // or PostPageComponent if it implements Resolve
}
```

**Comment by th0r** [comment]
@StevenLiekens static data can be merged into the dynamic one.

**Comment by StevenLiekens** [comment]
Then it's an interesting idea that would work for me except you still have that generic type parameter in `Resolve<T>`.

I wonder why Resolve is generic anyway? Why isn't it like this?

```ts
interface Resolve {
  resolve(route: ActivatedRouteSnapshot, state: RouterStateSnapshot): Observable<any> | Promise<any> | any
}
```

**Comment by th0r** [comment]
It can be generic with default type set to `any`.

**Comment by mrshelly** [comment]
??
Thank you , For read this Email !

________________________________
发件人: Yuriy Grunin <notifications@github.com>
发送时间: 2018年8月11日 16:11
收件人: angular/angular
抄送: Subscribed
主题: Re: [angular/angular] Resolve for side-effects only (#25428)


It can be generic with default type set to any.

―
